pallas-network/src/multiplexer.rs

```diff
@@ -198,85 +214,90 @@ pub enum Error {
+        let mut buf = vec![0u8; segment_size];
+        self.0.read_exact(&mut buf).await.map_err(Error::BearerIo)?;
+
+        Ok((header.protocol, buf))
     }
 
-    /// Peek the available data in search for a frame header
-    async fn peek_header(&mut self) -> Result<Header, Error> {
-        trace!("waiting for header buf");
-        self.cancellable_read(HEADER_LEN).await?;
+    fn demux(&mut self, protocol: Protocol, payload: Payload) -> Result<(), Error> {
+        if tracing::event_enabled!(tracing::Level::TRACE) {
+            trace!(protocol, data = hex::encode(&payload), "read from bearer");
+        }
+
+        self.1
+             .0
+            .send((protocol, payload))
+            .map_err(|err| Error::PlexerDemux(err.0 .0, err.0 .1))?;
+
+        Ok(())
+    }
 
-        trace!("found enough data for header");
-        let header = &self.1[..HEADER_LEN];
+    pub fn subscribe_recv(&self) -> tokio::sync::broadcast::Receiver<(Protocol, Payload)> {
+        self.1 .0.subscribe()
+    }
 
-        Ok(Header::from(header))
+    pub async fn tick(&mut self) -> Result<(), Error> {
+        let (protocol, payload) = self.read_segment().await?;
+        trace!(protocol, "demux happening");
+        self.demux(protocol, payload)
     }
 
-    // Cancel-safe read of a full segment from the bearer
-    pub async fn read_segment(&mut self) -> Result<(Protocol, Payload), Error> {
-        let header = self.peek_header().await?;
+    pub async fn run(&mut self) -> Result<(), Error> {
+        loop {
+            if let Err(err) = self.tick().await {
+                break Err(err);
+            }
+        }
+    }
+}
 
-        trace!("waiting for full segment buf");
-        let segment_size = HEADER_LEN + header.payload_len as usize;
+type Ingress = (
+    tokio::sync::mpsc::Sender<(Protocol, Payload)>,
+    tokio::sync::mpsc::Receiver<(Protocol, Payload)>,
+);
 
-        self.cancellable_read(segment_size).await?;
+type Clock = Instant;
 
-        trace!("draining segment buffer");
-        let segment = self.1.drain(..segment_size);
-        let payload = segment.skip(HEADER_LEN).collect();
+pub struct Muxer(BearerWriteHalf, Clock, Ingress);
 
-        Ok((header.protocol, payload))
+impl Muxer {
+    pub fn new(bearer: BearerWriteHalf) -> Self {
+        let ingress = tokio::sync::mpsc::channel(100); // TODO: define buffer
+        let clock = Instant::now();
+        Self(bearer, clock, ingress)
     }
 
-    pub async fn write_segment(
-        &mut self,
-        protocol: u16,
-        clock: &Instant,
-        payload: &[u8],
-    ) -> Result<(), std::io::Error> {
+    async fn write_segment(&mut self, protocol: u16, payload: &[u8]) -> Result<(), std::io::Error> {
         let header = Header {
             protocol,
-            timestamp: clock.elapsed().as_micros() as u32,
+            timestamp: self.1.elapsed().as_micros() as u32,
             payload_len: payload.len() as u16,
```

Comment:
We should check for integer overflow here. It's a protocol error if someone tries to write a segment larger than `65535` bytes.

Reply:
the `as 32` makes sure that we never overflow. If it goes higher than the u32::MAX, the value will hold the least significant bytes (it would look like a restart)